File: compiler/src/preprocessor/guards.cpp
#include "expression.hpp"
#include <algorithm>
#include <cmath>
#include <map>

namespace erlang_aot {
namespace {
using Values = std::vector<Value>;
using Guard = Value (*)(const Values &);
using Signature = std::pair<std::u32string_view, std::size_t>;

// Reject invalid Erlang types before indexing or operating on host containers.
const Value &typed(const Value &value, ValueKind kind) {
    if (value.kind != kind) {
        throw EvaluationFailure();
    }
    return value;
}

const Value &typed(Value &&, ValueKind) = delete;

bool proper_list(const Value &value) {
    return (value.kind == ValueKind::nil || value.kind == ValueKind::list) && !value.tail;
}

Value list_head(const Values &arguments) {
    const auto &value = typed(arguments[0], ValueKind::list);
    if (value.elements.empty()) {
        throw EvaluationFailure();
    }
    return value.elements.front();
}

Value list_tail(const Values &arguments) {
    const auto &value = typed(arguments[0], ValueKind::list);
    if (value.elements.empty()) {
        throw EvaluationFailure();
    }
    if (value.elements.size() == 1) {
        return value.tail ? *value.tail : Value{};
    }
    auto result = list({value.elements.begin() + 1, value.elements.end()});
    result.tail = value.tail;
    return result;
}

Value length(const Values &arguments) {
    if (!proper_list(arguments[0])) {
        throw EvaluationFailure();
    }
    return integer(arguments[0].elements.size());
}

Value element(const Values &arguments) {
    const auto &tuple = typed(arguments[1], ValueKind::tuple);
    const auto offset = index(arguments[0], tuple.elements.size());
    if (offset == 0) {
        throw EvaluationFailure();
    }
    return tuple.elements[offset - 1];
}

const Value *map_find(const Value &key, const Value &map) {
    typed(map, ValueKind::map);
    for (std::size_t i = 0; i < map.elements.size(); i += 2) {
        if (compare(key, map.elements[i], true) == 0) {
            return &map.elements[i + 1];
        }
    }
    return nullptr;
}

Value map_get(const Values &arguments) {
    const auto *value = map_find(arguments[0], arguments[1]);
    if (!value) {
        throw EvaluationFailure();
    }
    return *value;
}

Value size(const Values &arguments) {
    const auto &value = arguments[0];
    if (value.kind == ValueKind::tuple) {
        return integer(value.elements.size());
    }
    typed(value, ValueKind::bits);
    return integer(value.bits.size() / 8);
}

Value rounded(const Value &value, double (*operation)(double)) {
    if (value.kind == ValueKind::integer) {
        return value;
    }
    return integer(BigInt(operation(real(value))));
}

Value record(const Values &arguments) {
    if (arguments.size() < 2) {
        throw EvaluationFailure();
    }
    typed(arguments[1], ValueKind::atom);
    const auto &value = arguments[0];
    const auto count = arguments.size() == 3 ? index(arguments[2]) : value.elements.size();
    return boolean(value.kind == ValueKind::tuple && !value.elements.empty() && value.elements.size() == count &&
                   compare(value.elements[0], arguments[1], true) == 0);
}

Value binary_part(const Values &arguments) {
    const auto &binary = typed(arguments[0], ValueKind::bits);
    if (binary.bits.size() % 8 != 0) {
        throw EvaluationFailure();
    }
    auto position = arguments[1];
    auto count = arguments.back();
    if (arguments.size() == 2) {
        const auto &pair = typed(arguments[1], ValueKind::tuple);
        if (pair.elements.size() != 2) {
            throw EvaluationFailure();
        }
        count = pair.elements[1];
        position = pair.elements[0];
    }
    const auto total = binary.bits.size() / 8;
    const auto position_index = index(position, total);
    const bool backwards = integral(count) < 0;
    const auto length = index(backwards ? integer(-integral(count)) : count, total) * 8;
    const auto boundary = position_index * 8;
    if (length > (backwards ? boundary : binary.bits.size() - boundary)) {
        throw EvaluationFailure();
    }
    const auto start = backwards ? boundary - length : boundary;
    Value result;
    result.kind = ValueKind::bits;
    result.bits.assign(binary.bits.begin() + static_cast<std::ptrdiff_t>(start),
                       binary.bits.begin() + static_cast<std::ptrdiff_t>(start + length));
    return result;
}

// OTP's bounded integer predicate requires integer bounds as well as an integer candidate.
Value integer_range(const Values &a) {
    if (!std::ranges::all_of(a, [](const Value &value) { return value.kind == ValueKind::integer; })) {
        return boolean(false);
    }
    return boolean(compare(a[0], a[1]) >= 0 && compare(a[0], a[2]) <= 0);
}

// Any nonnegative integer arity is valid; only the matching external fun succeeds.
Value function_arity(const Values &arguments) {
    if (integral(arguments[1]) < 0) {
        throw EvaluationFailure();
    }
    return boolean(arguments[0].kind == ValueKind::function &&
                   compare(arguments[0].elements[1], arguments[1], true) == 0);
}

// This closed dispatch table is the pinned erl_internal:guard_bif/2 catalog.
const std::map<Signature, Guard> &guards() {
    static const std::map<Signature, Guard> table{
        {{U"abs", 1},
         [](const Values &a) {
             return a[0].kind == ValueKind::integer ? integer(a[0].integer < 0 ? -a[0].integer : a[0].integer)
                                                    : floating(std::abs(real(a[0])));
         }},
        {{U"binary_part", 2}, binary_part},
        {{U"binary_part", 3}, binary_part},
        {{U"bit_size", 1}, [](const Values &a) { return integer(typed(a[0], ValueKind::bits).bits.size()); }},
        {{U"byte_size", 1},
         [](const Values &a) { return integer((typed(a[0], ValueKind::bits).bits.size() + 7) / 8); }},
        {{U"ceil", 1}, [](const Values &a) { return rounded(a[0], std::ceil); }},
        {{U"floor", 1}, [](const Values &a) { return rounded(a[0], std::floor); }},
        {{U"round", 1}, [](const Values &a) { return rounded(a[0], std::round); }},
        {{U"trunc", 1}, [](const Values &a) { return rounded(a[0], std::trunc); }},
        {{U"float", 1}, [](const Values &a) { return floating(real(a[0])); }},
        {{U"element", 2}, element},
        {{U"hd", 1}, list_head},
        {{U"tl", 1}, list_tail},
        {{U"length", 1}, length},
        {{U"map_get", 2}, map_get},
        {{U"map_size", 1}, [](const Values &a) { return integer(typed(a[0], ValueKind::map).elements.size() / 2); }},
        {{U"is_map_key", 2}, [](const Values &a) { return boolean(map_find(a[0], a[1]) != nullptr); }},
        {{U"max", 2}, [](const Values &a) { return compare(a[0], a[1]) >= 0 ? a[0] : a[1]; }},
        {{U"min", 2}, [](const Values &a) { return compare(a[0], a[1]) <= 0 ? a[0] : a[1]; }},
        {{U"size", 1}, size},
        {{U"tuple_size", 1}, [](const Values &a) { return integer(typed(a[0], ValueKind::tuple).elements.size()); }},
        {{U"node", 0}, [](const Values &) { return atom(U"nonode@nohost"); }},
        {{U"node", 1},
         [](const Values &a) {
             typed(a[0], ValueKind::pid);
             return atom(U"nonode@nohost");
         }},
        {{U"self", 0},
         [](const Values &) {
             Value value;
             value.kind = ValueKind::pid;
             return value;
         }},
        {{U"is_integer", 3}, integer_range},
        {{U"is_record", 1}, record},
        {{U"is_record", 2}, record},
        {{U"is_record", 3}, record},
        {{U"is_function", 2}, function_arity}};
    return table;
}

// Type tests are independent of host runtime objects.
const std::map<std::u32string_view, bool (*)(const Value &)> &types() {
    static const std::map<std::u32string_view, bool (*)(const Value &)> table{
        {U"is_atom", [](const Value &v) { return v.kind == ValueKind::atom; }},
        {U"is_binary", [](const Value &v) { return v.kind == ValueKind::bits && v.bits.size() % 8 == 0; }},
        {U"is_bitstring", [](const Value &v) { return v.kind == ValueKind::bits; }},
        {U"is_boolean",
         [](const Value &v) { return v.kind == ValueKind::atom && (v.text == U"true" || v.text == U"false"); }},
        {U"is_float", [](const Value &v) { return v.kind == ValueKind::floating; }},
        {U"is_function", [](const Value &v) { return v.kind == ValueKind::function; }},
        {U"is_integer", [](const Value &v) { return v.kind == ValueKind::integer; }},
        {U"is_list", [](const Value &v) { return v.kind == ValueKind::list || v.kind == ValueKind::nil; }},
        {U"is_number", numeric},
        {U"is_map", [](const Value &v) { return v.kind == ValueKind::map; }},
        {U"is_pid", [](const Value &v) { return v.kind == ValueKind::pid; }},
        {U"is_port", [](const Value &v) { return v.kind == ValueKind::port; }},
        {U"is_reference", [](const Value &v) { return v.kind == ValueKind::reference; }},
        {U"is_tuple", [](const Value &v) { return v.kind == ValueKind::tuple; }}};
    return table;
}
} // namespace

bool guard_signature(std::u32string_view name, std::size_t arity) {
    return (arity == 1 && types().contains(name)) || guards().contains({name, arity});
}

Value guard_call(std::u32string_view name, const std::vector<Value> &arguments) {
    if (arguments.size() == 1) {
        if (const auto found = types().find(name); found != types().end()) {
            return boolean(found->second(arguments.front()));
        }
    }
    const auto found = guards().find({name, arguments.size()});
    if (found == guards().end()) {
        throw EvaluationFailure();
    }
    return found->second(arguments);
}
} // namespace erlang_aot

File: tests/fixtures/preprocessor/semantic/guards.erl
-module(guards).
-if(abs(-9) =:= 9).
selected(0,true).
-else.
selected(0,false).
-endif.
-if(abs(-1.5) =:= 1.5).
selected(1,true).
-else.
selected(1,false).
-endif.
-if(ceil(1.1) =:= 2).
selected(2,true).
-else.
selected(2,false).
-endif.
-if(floor(-1.1) =:= -2).
selected(3,true).
-else.
selected(3,false).
-endif.
-if(round(-1.5) =:= -2).
selected(4,true).
-else.
selected(4,false).
-endif.
-if(trunc(-1.9) =:= -1).
selected(5,true).
-else.
selected(5,false).
-endif.
-if(float(1) =:= 1.0).
selected(6,true).
-else.
selected(6,false).
-endif.
-if(length([1,2]) =:= 2).
selected(7,true).
-else.
selected(7,false).
-endif.
-if(hd([1|2]) =:= 1).
selected(8,true).
-else.
selected(8,false).
-endif.
-if(tl([1|2]) =:= 2).
selected(9,true).
-else.
selected(9,false).
-endif.
-if(is_list([1|2])).
selected(10,true).
-else.
selected(10,false).
-endif.
-if(size({a,b}) =:= 2).
selected(11,true).
-else.
selected(11,false).
-endif.
-if(tuple_size({}) =:= 0).
selected(12,true).
-else.
selected(12,false).
-endif.
-if(element(1,{a}) =:= a).
selected(13,true).
-else.
selected(13,false).
-endif.
-if(map_get(1,#{1=>a,1.0=>b}) =:= a).
selected(14,true).
-else.
selected(14,false).
-endif.
-if(map_size(#{1=>a,1.0=>b}) =:= 2).
selected(15,true).
-else.
selected(15,false).
-endif.
-if(is_map_key(a,#{a=>1})).
selected(16,true).
-else.
selected(16,false).
-endif.
-if(#{a=>1}#{a:=2} =:= #{a=>2}).
selected(17,true).
-else.
selected(17,false).
-endif.
-if(bit_size(<<3:5>>) =:= 5).
selected(18,true).
-else.
selected(18,false).
-endif.
-if(byte_size(<<3:5>>) =:= 1).
selected(19,true).
-else.
selected(19,false).
-endif.
-if(size(<<3:5>>) =:= 0).
selected(20,true).
-else.
selected(20,false).
-endif.
-if(binary_part(<<1,2,3>>, {1,2}) =:= <<2,3>>).
selected(21,true).
-else.
selected(21,false).
-endif.
-if(binary_part(<<1,2,3>>, 3,-2) =:= <<2,3>>).
selected(22,true).
-else.
selected(22,false).
-endif.
-if(<<16#123:12/little>> =:= <<16#23,1:4>>).
selected(23,true).
-else.
selected(23,false).
-endif.
-if(<<65/utf8>> =:= <<65>>).
selected(24,true).
-else.
selected(24,false).
-endif.
-if(<<16#10000/utf16>> =:= <<16#d800:16,16#dc00:16>>).
selected(25,true).
-else.
selected(25,false).
-endif.
-if(<<1.0:16/float>> =:= <<16#3c00:16>>).
selected(26,true).
-else.
selected(26,false).
-endif.
-if(<<1.0:32/float>> =:= <<16#3f800000:32>>).
selected(27,true).
-else.
selected(27,false).
-endif.
-if(is_atom(a)).
selected(28,true).
-else.
selected(28,false).
-endif.
-if(is_boolean(false)).
selected(29,true).
-else.
selected(29,false).
-endif.
-if(is_binary(<<1>>)).
selected(30,true).
-else.
selected(30,false).
-endif.
-if(is_bitstring(<<1:1>>)).
selected(31,true).
-else.
selected(31,false).
-endif.
-if(is_float(1.0)).
selected(32,true).
-else.
selected(32,false).
-endif.
-if(is_integer(1)).
selected(33,true).
-else.
selected(33,false).
-endif.
-if(is_integer(3,1,4)).
selected(34,true).
-else.
selected(34,false).
-endif.
-if(is_integer(3,1.0,4.0)).
selected(35,true).
-else.
selected(35,false).
-endif.
-if(is_integer(3,a,z)).
selected(36,true).
-else.
selected(36,false).
-endif.
-if(is_map(#{})).
selected(37,true).
-else.
selected(37,false).
-endif.
-if(is_number(1.0)).
selected(38,true).
-else.
selected(38,false).
-endif.
-if(is_pid(self())).
selected(39,true).
-else.
selected(39,false).
-endif.
-if(is_port(self())).
selected(40,true).
-else.
selected(40,false).
-endif.
-if(is_reference(self())).
selected(41,true).
-else.
selected(41,false).
-endif.
-if(is_tuple({})).
selected(42,true).
-else.
selected(42,false).
-endif.
-if(is_function(a)).
selected(43,true).
-else.
selected(43,false).
-endif.
-if(is_function(a,256)).
selected(44,true).
-else.
selected(44,false).
-endif.
-if(not is_function(a,256)).
selected(45,true).
-else.
selected(45,false).
-endif.
-if(is_record({r,a},r,2)).
selected(46,true).
-else.
selected(46,false).
-endif.
-if(is_record({r,a},r)).
selected(47,true).
-else.
selected(47,false).
-endif.
-if(is_record({r,a})).
selected(48,true).
-else.
selected(48,false).
-endif.
-if(is_record({r,a},r,0)).
selected(49,true).
-else.
selected(49,false).
-endif.
-if(min(1,1.0) =:= 1).
selected(50,true).
-else.
selected(50,false).
-endif.
-if(max(1,1.0) =:= 1).
selected(51,true).
-else.
selected(51,false).
-endif.
-if(node(self()) =:= node()).
selected(52,true).
-else.
selected(52,false).
-endif.
-if(erlang:'+'(1,2) =:= 3).
selected(53,true).
-else.
selected(53,false).
-endif.
-if(erlang:'not'(false)).
selected(54,true).
-else.
selected(54,false).
-endif.
-if('+'(1,2) =:= 3).
selected(55,true).
-else.
selected(55,false).
-endif.
-if(erlang:defined(MODULE)).
selected(56,true).
-else.
selected(56,false).
-endif.
-if((1 bsl 80)+1 > (1 bsl 80)).
selected(57,true).
-else.
selected(57,false).
-endif.
-if(-7 div 3 =:= -2).
selected(58,true).
-else.
selected(58,false).
-endif.
-if(-7 rem 3 =:= -1).
selected(59,true).
-else.
selected(59,false).
-endif.
-if((1 bsl -2) =:= 0).
selected(60,true).
-else.
selected(60,false).
-endif.
-if((-7 bsr 2) =:= -2).
selected(61,true).
-else.
selected(61,false).
-endif.
-if((bnot 3) =:= -4).
selected(62,true).
-else.
selected(62,false).
-endif.
-if(1 == 1.0).
selected(63,true).
-else.
selected(63,false).
-endif.
-if(1 =:= 1.0).
selected(64,true).
-else.
selected(64,false).
-endif.
-if({1} == {1.0}).
selected(65,true).
-else.
selected(65,false).
-endif.
-if([1] == [1.0]).
selected(66,true).
-else.
selected(66,false).
-endif.
-if(#{1=>x} == #{1.0=>x}).
selected(67,true).
-else.
selected(67,false).
-endif.
-if(#{a=>1} == #{a=>1.0}).
selected(68,true).
-else.
selected(68,false).
-endif.
-if([] < [0]).
selected(69,true).
-else.
selected(69,false).
-endif.
-if({a} < {0,0}).
selected(70,true).
-else.
selected(70,false).
-endif.
-if(9007199254740993 > 9007199254740992.0).
selected(71,true).
-else.
selected(71,false).
-endif.
-if(1 < 2 < 3).
selected(72,true).
-else.
selected(72,false).
-endif.
-if(true orelse missing()).
selected(73,true).
-else.
selected(73,false).
-endif.
-if(true orelse <<1/unknown>>).
selected(74,true).
-else.
selected(74,false).
-endif.
-if(true orelse X).
selected(75,true).
-else.
selected(75,false).
-endif.
-if(true orelse (1 div 0)).
selected(76,true).
-else.
selected(76,false).
-endif.
-if(is_tuple({r,1})).
selected(77,true).
-else.
selected(77,false).
-endif.
-if(true orelse #r.a).
selected(78,true).
-else.
selected(78,false).
-endif.
-if(true orelse X#r.a).
selected(79,true).
-else.
selected(79,false).
-endif.
-if(true orelse X#r.a).
selected(80,true).
-else.
selected(80,false).
-endif.
-if(~s"abc" =:= "abc").
selected(81,true).
-else.
selected(81,false).
-endif.
-if(~b"å" =:= <<195,165>>).
selected(82,true).
-else.
selected(82,false).
-endif.
-if(~"abc" =:= <<97,98,99>>).
selected(83,true).
-else.
selected(83,false).
-endif.
-if(~s"a"bad =:= "a").
selected(84,true).
-else.
selected(84,false).
-endif.
-if(is_function(fun erlang:length/1)).
selected(85,true).
-else.
selected(85,false).
-endif.
-if(<<1:8/unit:2>> =:= <<0,1>>).
selected(86,true).
-else.
selected(86,false).
-endif.
-if(<<1/integer-float>> =:= <<1>>).
selected(87,true).
-else.
selected(87,false).
-endif.
-if(<<1:2/utf8>> =:= <<1>>).
selected(88,true).
-else.
selected(88,false).
-endif.
-if(true orelse <<1:2/utf8>>).
selected(89,true).
-else.
selected(89,false).
-endif.
-if(true orelse <<1/unit:0>>).
selected(90,true).
-else.
selected(90,false).
-endif.
-if(#{2=>a} < #{1.0=>a}). map_key_order(true). -else. map_key_order(false). -endif.
-if(<<[]>> =:= <<>>). empty_segment(true). -else. empty_segment(false). -endif.
-if(1 / (1 bsl 2000) =:= 0.0). float_overflow(true). -else. float_overflow(false). -endif.
